app/services/fact_cleaner.rb

```diff
@@ -39,11 +43,38 @@ def find_compose_orphaned_facts(found)
   end
 
   def delete!(to_delete)
-    logger.info "#{to_delete.size} facts will be removed"
+    logger.debug "#{to_delete.size} facts will be removed"
     to_delete.each do |fact|
       logger.debug { "Removing fact #{fact}" }
       fact.destroy
     end
-    logger.info "Cleanup of facts finished!"
+    logger.debug "Cleanup of facts finished!"
+  end
+
+  def find_excluded_facts
+    logger.debug "Searching for facts that match excluded pattern..."
+    fact_name_ids = []
+
+    excluded_facts_regex = FactImporter.new(nil, {}).send(:excluded_facts)
+
+    FactName.unscoped.find_in_batches(:batch_size => 100) do |names_batch|
+      names_batch.select! { |fact_name| fact_name.name.match(excluded_facts_regex) }
+      fact_name_ids += names_batch.map(&:id)
+    end
+
+    fact_name_ids.uniq
+  end
+
+  def delete_excluded_facts
+    fact_names = find_excluded_facts
+
+    # delete related fact values first
+    logger.debug "Removing values of excluded facts..."
+    deleted_values = FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
+    logger.debug "Removed #{deleted_values} associated fact value records."
+    logger.debug "Removing excluded fact names..."
+    deleted_names = FactName.unscoped.where(:id => fact_names).delete_all
+    logger.debug "Removed #{deleted_names} fact name records."
+    deleted_names
```

Comment:
I am usually against debug messages like "I am doing something..." the "Removed" lines are enough to tell you it has been processed.

Reply:
It really helps to identify slow queries - you will see the "starting.." message, but not the "finished" one.

Reply:
Fair enough, although AR time is reported for each request by Rails framework and then there is SQL server tracking, but I buy that. We miss TRACE level so much.